### PR TITLE
[cli] Add support for environments

### PR DESCRIPTION
**Pull request by pgavlin**
These changes add support for ESC environments to the Pulumi CLI. This involves two major changes:

- Support for the `env` subcommand
- Support for the `environment` stanza in stack config files

The former reuses the command from `esc` itself with a little rebranding.

The latter adds support to stack config files for an `environment` property of the form:

```yaml
environment:
  - list
  - of
  - environment
  - names
```

If this property is present in a stack's config file, the CLI will open the and merge the listed environments during `pulumi up` et. al. If an object-valued `pulumiConfig` property is present in the opened environment, its values will be merged on top of the stack's config prior to whatever operation is to be performed. If an object-valued `environmentVariables` property is present inthe opened environment, its values will be published as environment variables prior to the Pulumi operation. Any values in the open environment's `pulumiConfig` or `environmentVariables` that are marked as secret will be encrypted in the resulting config and will be filtered from the command's logs.

### CODE REVIEW

File: pkg/go.mod [hunk 1]
```diff
@@ -1,6 +1,6 @@
 module github.com/pulumi/pulumi/pkg/v3
 
-go 1.20
+go 1.21.0
```
Comment (Frassle) [inline pkg/go.mod]: It's not great to bump to 1.21 here, providers still have build time dependencies on Go 1.20 and our official support position has been well work with 1.20 and 1.21. If this is hard to revert we'll just have to bite it and hope the fallouts ok.

File: sdk/go/common/apitype/environments.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apitype
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pulumi/esc"
```
Comment (Frassle) [inline sdk/go/common/apitype/environments.go:21]: If we're taking a build time dependency on pulumi/esc is that now a (or soon going to be) a stable API where we'll endeavour not to make break changes?

File: pkg/backend/httpstate/environments.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package httpstate
```
Comment (justinvp) [inline pkg/backend/httpstate/environments.go:1]: Nit: Copyright header

File: pkg/cmd/pulumi/about_env.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
+// Copyright 2016-2022, Pulumi Corporation.
```
Comment (justinvp) [inline pkg/cmd/pulumi/about_env.go:1]:
```suggestion
// Copyright 2016-2023, Pulumi Corporation.
```

**Review by Frassle — COMMENTED** [review]
There's not any testing to go with this? I assume we'll be able to add integration tests against the esc service with something like how we have moolumi for cloud integration tests?

File: sdk/go/pulumi-language-go/go.mod [hunk 5]
```diff
@@ -11,10 +11,10 @@ require (
 	github.com/blang/semver v3.5.1+incompatible
 	github.com/golang/protobuf v1.5.3
 	github.com/opentracing/opentracing-go v1.2.0
-	github.com/pulumi/pulumi/pkg/v3 v3.0.0-00010101000000-000000000000
+	github.com/pulumi/pulumi/pkg/v3 v3.78.2-0.20230926151835-2b44cf6ec1be
```
Comment (justinvp) [inline sdk/go/pulumi-language-go/go.mod:14]: Could this be `v3.87.0`?

File: sdk/nodejs/cmd/pulumi-language-nodejs/go.mod [hunk 6]
```diff
@@ -14,9 +14,9 @@ require (
 	github.com/google/shlex v0.0.0-20191202100458-e7afc7fbc510
 	github.com/hashicorp/go-multierror v1.1.1
 	github.com/opentracing/opentracing-go v1.2.0
-	github.com/pulumi/pulumi/pkg/v3 v3.0.0-00010101000000-000000000000
+	github.com/pulumi/pulumi/pkg/v3 v3.78.2-0.20230926151835-2b44cf6ec1be
```
Comment (justinvp) [inline sdk/nodejs/cmd/pulumi-language-nodejs/go.mod:17]: Could this be `v3.87.0`?

File: sdk/python/cmd/pulumi-language-python/go.mod [hunk 7]
```diff
@@ -10,9 +10,9 @@ replace (
 require (
 	github.com/blang/semver v3.5.1+incompatible
 	github.com/golang/protobuf v1.5.3
-	github.com/pulumi/pulumi/pkg/v3 v3.0.0-00010101000000-000000000000
+	github.com/pulumi/pulumi/pkg/v3 v3.78.2-0.20230926151835-2b44cf6ec1be
```
Comment (justinvp) [inline sdk/python/cmd/pulumi-language-python/go.mod:13]: Could this be `v3.87.0`?

File: tests/go.mod [hunk 8]
```diff
@@ -13,9 +13,9 @@ require (
 	github.com/golang/protobuf v1.5.3
 	github.com/grapl-security/pulumi-hcp/sdk v0.1.14
 	github.com/hexops/autogold v1.3.0
-	github.com/pulumi/pulumi/pkg/v3 v3.0.0-00010101000000-000000000000
+	github.com/pulumi/pulumi/pkg/v3 v3.78.2-0.20230926151835-2b44cf6ec1be
```
Comment (justinvp) [inline tests/go.mod:16]: Could this be `v3.87.0`?

File: tests/integration/construct_component_provider_propagation/go/go.mod [hunk 9]
```diff
@@ -1,17 +1,17 @@
 module github.com/pulumi/pulumi/tests/construct_component_provider_propagation
 
-go 1.18
+go 1.20
 
 replace github.com/pulumi/pulumi/sdk/v3 => ../../../../sdk
 
-require github.com/pulumi/pulumi/sdk/v3 v3.62.0
+require github.com/pulumi/pulumi/sdk/v3 v3.85.1-0.20230926151835-2b44cf6ec1be
```
Comment (justinvp) [inline tests/integration/construct_component_provider_propagation/go/go.mod:7]: Could this be `v3.87.0`?

File: tests/integration/construct_component_resource_options/go/go.mod [hunk 10]
```diff
@@ -1,17 +1,17 @@
 module github.com/pulumi/pulumi/tests/construct_component_resource_options
 
-go 1.18
+go 1.20
 
 replace github.com/pulumi/pulumi/sdk/v3 => ../../../../sdk
 
-require github.com/pulumi/pulumi/sdk/v3 v3.62.0
+require github.com/pulumi/pulumi/sdk/v3 v3.85.1-0.20230926151835-2b44cf6ec1be
```
Comment (justinvp) [inline tests/integration/construct_component_resource_options/go/go.mod:7]: Could this be `v3.87.0`?

File: tests/integration/construct_nested_component/go/go.mod [hunk 11]
```diff
@@ -1,8 +1,8 @@
 module github.com/pulumi/pulumi/tests/construct_nested_component
 
-go 1.18
+go 1.20
 
-require github.com/pulumi/pulumi/sdk/v3 v3.40.1
+require github.com/pulumi/pulumi/sdk/v3 v3.85.1-0.20230926151835-2b44cf6ec1be
```
Comment (justinvp) [inline tests/integration/construct_nested_component/go/go.mod:5]: Could this be `v3.87.0`?

**Review by justinvp — APPROVED** [review]
LGTM otherwise

File: tests/integration/construct_component_provider_propagation/go/go.mod (hunk 9 above)
Comment (pgavlin) [reply tests/integration/construct_component_provider_propagation/go/go.mod:7]: This is from `make tidy` ¯\_(ツ)_/¯

File: tests/integration/construct_component_resource_options/go/go.mod (hunk 10 above)
Comment (pgavlin) [reply tests/integration/construct_component_resource_options/go/go.mod:7]: Same--this is `make tidy`'s doing--I'm inclined to take what it gives us

File: tests/integration/construct_nested_component/go/go.mod (hunk 11 above)
Comment (pgavlin) [reply tests/integration/construct_nested_component/go/go.mod:5]: See above

**Comment by pgavlin** [comment]
> There's not any testing to go with this? I assume we'll be able to add integration tests against the esc service with something like how we have moolumi for cloud integration tests?

I'm working on adding tests right now. Most of the functionality is covered by tests in `esc` itself.

File: sdk/go/common/apitype/environments.go (hunk 2 above)
Comment (pgavlin) [reply sdk/go/common/apitype/environments.go:21]: Yes. `esc`'s public API is intended to be stable.

**Comment by EronWright** [comment]
Noting that the values are merged rather than replaced, unlike how project-level config is handled. See https://github.com/pulumi/pulumi/issues/13176

**Pull request merged**